### ISSUE TITLE
Nunchuck not working

### ISSUE THREAD

**Issue by WilliamAnderssonDev** [EFeru/hoverboard-firmware-hack-FOC#206]
### Your System
- VARIANT: Nunchuck
- CONTROL TYPE: 
- CONTROL MODE:
- What is your application?
    - Golf Trolley

Hello, i have connected a Nunchuck like described in https://github.com/EmanuelFeru/hoverboard-firmware-hack-FOC/wiki/Variant-NUNCHUK 

It beeps like normal when i start and i have no error beep code but when i move the nunchuck nothing happens and its official Wii nunchuck

**Comment by Candas1** [comment]
Hi,

Have you maybe wrongly entered auto-calibration mode and didn't proceed properly ?

**Comment by WilliamAnderssonDev** [comment]
How do i do that?

**Comment by Candas1** [comment]
Did you read the wiki ?
https://github.com/EmanuelFeru/hoverboard-firmware-hack-FOC/wiki/Variant-Nunchuk

**Comment by WilliamAnderssonDev** [comment]
Yes i have, i tried to autocalibrate now and still nothing

**Comment by Candas1** [comment]
No error beeps ?
How is your mainboard looking like ?

**Comment by WilliamAnderssonDev** [comment]
Please check my video here https://www.youtube.com/watch?v=jyXJUdMHx0g

**Comment by Candas1** [comment]
Can you also record on video how you do the auto-calibration please ?

**Comment by WilliamAnderssonDev** [comment]
Ok, i did flash erase to remove auto calibration, should i calibrate again?

**Comment by Candas1** [comment]
Auto-calibration is nice to have with nunchuk to have joystick properly centered, but it's not mandatory.
Nunchuk variant should work out of the box even without auto-calibration if you have proper wiring.
I tried it myself many times.

So it's still no working after erasing and flashing ?

**Comment by WilliamAnderssonDev** [comment]
Nope the video is my latest step

**Comment by WilliamAnderssonDev** [comment]
Nunchuck SDA(green) is connected to hoverboard green. Nunchuck red is to 3.3v, Nunchuck white is to hoverboard black and nunchuck SCL(yellow) is to board blue. And i have resistor like said. I connected everything to the right sideboard as described in this picture. https://raw.githubusercontent.com/EmanuelFeru/hoverboard-firmware-hack-FOC/master/docs/pictures/mainboard_pinout.png

**Comment by Candas1** [comment]
Are you sure you picked the right variant as mentioned [here](https://github.com/EmanuelFeru/hoverboard-firmware-hack-FOC/wiki/Variants) ?
How are you compiling and flashing ?

**Comment by WilliamAnderssonDev** [comment]
Yep, im using platformio.
platformio.ini: https://pastebin.com/DtYrCHDE
config.h: https://pastebin.com/VqPWQwGL

**Comment by WilliamAnderssonDev** [comment]
i use st link v2 to flash with STM32CubeProgrammer

**Comment by WilliamAnderssonDev** [comment]
I have tried the arduino with USART too but i just get the 3 beeps so i dont understand what im doing wrong

**Comment by Candas1** [comment]
Maybe you have problems with the MCU ports on the right cable which could also explain why nunchuk is not working, you can try USART2 and left cable
https://github.com/EmanuelFeru/hoverboard-firmware-hack-FOC/wiki/Variant-USART

**Comment by WilliamAnderssonDev** [comment]
Doesnt arduino destroy usart2 as it outputs 5v?

**Comment by Candas1** [comment]
only usart3 ones are 5v tolerant
[image](https://github.com/EmanuelFeru/hoverboard-firmware-hack-FOC/blob/master/docs/pictures/mainboard_pinout.png)

**Comment by WilliamAnderssonDev** [comment]
Then i can't try usart2 if i cant get nunchuck there. I dont know why nothing works for me because the hoverboard was working when i took it apart.

Can you help me with config.h and platformio.ini for nunchuck use like you had
and maybe one for arduino so im sure that its not my config thats wrong

**Comment by Candas1** [comment]
Well apart from selecting the right variant, there shouldn't be anything to change
Do you have a ftdi to check in ascii debug what input values look like?

**Comment by WilliamAnderssonDev** [comment]
I dont have one. I just dont understand why nothing works because it was a fully functional hoverboard

**Comment by Candas1** [comment]
Maybe you are doing something wrong, or maybe your board is different.

As you can see firmware works
https://youtu.be/SYkKr-bOYis

**Comment by WilliamAnderssonDev** [comment]
My board is same as yours

**Comment by vamfun** [comment]
You can try usart2  with a voltage divider.   Put 5 v across pot and adjust center to 3.3 v .   Lock pot with tape and then put Arduino across pot and connect to usart2 .

**Comment by 1a2s3d4f5TDS** [comment]
Hi everybody i have a same situation as topic starter
No beeps no spin. when i try disconnect signal wire blue or green i got error beeps. One more important thing. My board variant 1 (usual 0)

**Comment by SerjeoWN** [comment]
> ### Your System
> * VARIANT: Nunchuck
> * CONTROL TYPE:
> * CONTROL MODE:
> * What is your application?
>   
>   * Golf Trolley
> 
> Hello, i have connected a Nunchuck like described in https://github.com/EmanuelFeru/hoverboard-firmware-hack-FOC/wiki/Variant-NUNCHUK
> 
> It beeps like normal when i start and i have no error beep code but when i move the nunchuck nothing happens and its official Wii nunchuck

Hi everybody i have a same situation as topic starter
No beeps no spin. when i try disconnect signal wire blue or green i got error beeps. One more important thing. My board variant 1 (usual 0)

**Comment by SerjeoWN** [comment]
Hi everybody i have a same situation as topic starter
No beeps no spin. when i try disconnect signal wire blue or green i got error beeps. One more important thing. My board variant 1 (usual 0)

**Comment by Candas1** [comment]
You probably did a bad calibration by mistake, either do a proper calibration or erase/flash the chip again. Check the details in the wiki.

**Comment by SerjeoWN** [comment]
> You probably did a bad calibration by mistake, either do a proper calibration or erase/flash the chip again. Check the details in the wiki.

[https://www.youtube.com/watch?v=5dqO9VF2zsc](url)

1) no second sound after turn on . only after when im unplug nunchuck 2) when i play with Nunchuck plug , i can run motor , but until next off/on

**Comment by Candas1** [comment]
Have you calibrated as I mentioned?
Motors won't enable at startup if the inputs are not very small. So if your joystick is not well centered it won't work. That's why you need to calibrate the inouts either manually with debug serial or the auto calibration.

**Comment by SerjeoWN** [comment]
> Have you calibrated as I mentioned?

> 

https://github.com/EFeru/hoverboard-firmware-hack-FOC/wiki/Input-Calibration
yes
Auto-calibration finds the minimum, middle and maximum values for the inputs. It also identifies the type of input and checks if it can be protected (ADC only).

Turn the board on and press the power button for more than 5 sec and release after the beep sound
Move the potis freely to the min and max limits repeatedly
Release potis to the resting position
Press the power button to confirm or wait for the 20 sec timeout


> Also as mentioned in the wiki you should better use pull up resistors.

yes. 91 kOm Blue-Groud and Green-Ground


i think i catch my issue
Please check my video 
https://www.youtube.com/watch?v=nKbINWaS1-o

**Comment by Candas1** [comment]
I will try to reproduce it at home tomorrow

**Comment by Candas1** [comment]
OK, I found a very weird bug, and it confirms what @WilliamAnderssonDev, sorry I didn't take it seriously back then.
I think it might not be happening on all boards, and the board I used now has the same issue. Because I had tested nunchuk in the past and there was no issue.

@SerjeoWN, can you please try following change, and if confirmed I will do a pull request.
[Here](https://github.com/EFeru/hoverboard-firmware-hack-FOC/blob/4770b13b2cfc3a952f17f6217ec254639d10a0f5/Src/util.c#L293) in your util.c file, please add the following command `HAL_Delay(100);`

Before:
```
I2C_Init();
Nunchuk_Init();
```

After:
```
I2C_Init();
HAL_Delay(100);
Nunchuk_Init();
```

**Comment by SerjeoWN** [comment]
> I found a very weird bug, and it confirms what @WilliamAnderssonDev, sorry I didn't take it seriously back then.

Thank you. It seemed to me that I was going crazy and losing faith in my knowledge and the power of the mind.

> I think it might not be happening on all boards, and the board I used now has the same issue.

Does your board have an alternative pin configuration? 1 not 0 ?
maybe this is a problem alternative boards ?

> can you please try following change, and if confirmed I will do a pull request.

I'll test it for a few days and let you know the result.

**Comment by Candas1** [comment]
No it's not an alternative board.
I am not sure I am confused now. This delay fixed the issue for me. Looks like we try to initialize the nunchuk too quickly after the initialization of I2C.
But there is a piece of code that is supposed to initialize the nunchuk again after few failures. Maybe we broke that piece of code during last change in december 2020... I would need to try older versions of the firmware to be sure.

**Comment by SerjeoWN** [comment]
> like we try to initialize the nunchuk too quickly after the initialization of I2C.
I think so too.
1)The nunchuk gets power quickly.
2)But the hover board hasn't booted yet
3)Nunchuk does not receive a response from the board, and goes to sleep
4)The board has loaded and does not find the nunchuck (he is sleeping)
5)error.

**Comment by Candas1** [comment]
The weird thing is it's always working after flashing.
So if I try debugging from platformio, I cannot reproduce it as it's first flashing the firmware.
I will do more tests in the evening.

In the meantime please let me know if the workaround fixed the issue for you.

**Comment by SerjeoWN** [comment]
> The weird thing is it's always working after flashing.

not always. Often I have to upload another firmware. I suspect that the flash memory is not completely erased, even when I manually erase the memory through the STM software.

**Comment by Candas1** [comment]
This is only when calibration is bad and saved to eeprom.
But I checked it, it happens even without the calibration in eeprom.

**Comment by Candas1** [comment]
I tried a version from October 2020, still couldn't get Nunchuk to work. Almost seems like it was never working, although I have recorded videos where it's working.
Only this delay helps.

I am wondering now if it could be due to an update in the toolchain, resulting in a different binary.

**Comment by SerjeoWN** [comment]
> In the meantime please let me know if the workaround fixed the issue for you.

yes and no.

here's some more video with my terrible english.  ))))
https://www.youtube.com/watch?v=BZu9w7EcN20&t=24s
and looks like i found another bug
https://www.youtube.com/watch?v=nKbINWaS1-o&t=5s

**Comment by Candas1** [comment]
From your video, I cannot understand what is not working.
Can you rephrase it here please ?

**Comment by SerjeoWN** [comment]
> From your video, I cannot understand what is not working. Can you rephrase it here please ?

after the first turn on, the Nunchuk  does not initialize, there is no second tune, until you start moving the joystick.

**Comment by Candas1** [comment]
But is it without calibration ? 
Can you check the debug serial output with a ftdi ?

There is a security feature that will enable wheels (second tune) only if the values are really small (-50 to 50).
If you joystick is a bit off center it can happen.

**Comment by sdegeorgio** [comment]
I've also had big problems getting the nunchuck to work.  I've studied the code and made quite a few changes to improve the connectivity.  It took lots of scoping on the I2C bus to ensure reliable bus re-starts in event of errors...  I hope it may help improve the experience for others too.
Stephan.

**Comment by Candas1** [comment]
Hi Stephan,
I tried your changes now, it definitely does a much better job when unplugging/plugging the nunchuk. 
I also tried Dual inputs with USART and NUNCHUK, it worked well. 

Thank for the great contribution.
Adding @EFeru for visibility has he his the owner of the repository.

**Comment by Candas1** [comment]
@SerjeoWN, have you tried it as well? Any feedback?

**Comment by EFeru** [comment]
Should be fixed by #260
